### ISSUE TITLE
Incorporate Google Sign-In

### ISSUE THREAD

**Issue by namoscato** [bromagosa/snapCloud#111]
Related to jmoenig/Snap--Build-Your-Own-Blocks#1805, I would like to suggest adding support for [Google Sign-In](https://developers.google.com/identity/sign-in/web/). I am a volunteer co-teacher for [TEALS](https://www.tealsk12.org/), and the students at my school are familiar with signing into Google products. Incorporating Google Sign-In would remove the need for them to remember another set of credentials.

I am slowly getting up to speed with this project and the current state of things (i.e. it looks like production Snap is hitting up this API, but there is a beta [social website](https://github.com/bromagosa/SnapSite)?). Happy to work through implementation specifics if we think this is a worthwhile initiative!

**Comment by bromagosa** [comment]
Other people have suggested Google Sign-In, but this involves politics and needs a lengthy discussion on convenience vs privacy. We don't store any personal information at all, while Google stores even your foot size. I have strong reservations on using proprietary and multi-billion dollar company owned services in educational software, however convenient it may be to just click on a button and be logged in. I'm pretty sure @brianharvey will have insightful thoughts to add to the discussion.

If something like Mozilla Persona had taken off, that'd be the correct thing to support. With Google, though, we can't really know what they'd be doing with the loads of personal information they'd collect from our students.

> i.e. it looks like production Snap is hitting up this API, but there is a beta social website?

Yes, production Snap! is hitting this API since February this year. There's a beta social website that we'll eventually finish. We have no funding for it at the moment so things got a bit slow, but we'll get there.

**Comment by namoscato** [comment]
Thanks for the feedback, @bromagosa!

I hear you with privacy concerns, but Snap! would not have to store any of that information, nor would this feature force users to start using Google. It would still be on the user to determine how they would manage their account.

At any rate, I would be happy to help with this initiative, so keep me posted!

**Comment by namoscato** [comment]
@brianharvey, I do not believe IP addresses would be inherently concealed as we would be redirecting to Google on the client and invoking their API on the backend (to verify a token upon login).

**Comment by brianharvey** [comment]
Bummer.  Can we just send them Berkeley's IP address instead?

**Comment by bromagosa** [comment]
> So it'll make a lot of teachers happy if we do this.

Yep. As I said, it's all about convenience vs privacy ;)

Selling your soul to the devil has its advantages.

**Comment by namoscato** [comment]
Per [their documentation](https://developers.google.com/identity/sign-in/web/sign-in), here is how the flow works after Google JavaScript and a button are added to the page:

1. User clicks on button and is redirected to Google. At this point, we would intentionally _not_ have control over the user's Google sign in process, which is why we cannot conceal user IPs. With any single sign-on provider, one must trust their integrity. In this step, we are assuming that if a user is able to authenticate with Google, they are in fact who they say they are. Hopefully this assumption is already the case for Snap! users with a Gmail address as a compromise to their Gmail account would inherently compromise their Snap! account as well (given that signup / password reset actions send email).
2. Upon successful Google authentication, Google invokes a JavaScript callback on our client with an ID token.
3. Our client invokes an endpoint on our server with this token.
4. Our server verifies the integrity of the token by invoking a [Google `tokeninfo` endpoint](https://developers.google.com/identity/sign-in/web/backend-auth#calling-the-tokeninfo-endpoint). (Google does not have a client library for Lua, but verifying the token's integrity over HTTPS is certainly a valid alternative.)
5. We would compare the email associated with the Google token with a user in Snap!, create a new Snap! record for the user (if they do not already exist), and establish a session.

I see where you guys are coming from on the privacy front. In most cases, I imagine Google Sign-In will be used by users who already have a Google account which they are using in school. In those contexts, I do not see how we would be introducing any new privacy concerns.

**Comment by brianharvey** [comment]
How do we associate the user's Snap<i>!</i> account name with their Google account name?  On Scratch, kids are encouraged to have user names that _aren't_ the same as their email name.  I guess that before they can use Google login, they have to register their Google account name in our database, right?  So we have to keep an important piece of identifying information.  But I guess if they're over 13 we already know their email address... 

Maybe the right solution is to offer Google login but with a box around the button saying something like "If your school uses Google for Education, you can set up Snap<i>!</i> to accept your school ID..."  In that situation the cat is already out of the bag.

FWIW, I think Google is the least evil of the evil Internet companies.  Most evil are the ISPs (exvept mine, Socic, which commits itself to a very strict privacy policy) because you can't opt out and they read everything.  Then Facebook.  Then everbofy else, and then Google, which gets points because they make pretty readable promises, if not strict ones, and then keep them.  Except, it turns out, with respect to the NSA, but they're no worse than other companies about that and better than some.

Let me put it this way.  I'm certainly not going to use Google authentication for my account.  But our meta-principle should be not to withhold things our users want to keep ourselves pure,  (Except if they want Facebook login.  Facebook are truly Snidely-Whiplash-level evil.

**Comment by namoscato** [comment]
I think there are some options. We could (1) use the user's full Google email or (2) local-part as their username, but we might have to handle conflicts (i.e. Snap! accounts that already exist with that username).

Are email addresses unique across Snap! accounts?

Regardless, I think we would have to prompt new Google Sign-In users with the ability to set their birth date to complete the signup flow as that data will not be exposed via their Google profile.

**Comment by brianharvey** [comment]
Oh foo, you mean kids would have to use their Google account as their Snap<i>!</i> user name?  This is the straw that breaks the camel's back, for me, I think.

**Comment by cycomachead** [comment]
In my view: 
Users or schools would be at their sole discretion to use Google logins. We'd be enabling you to give up some data to Google, but we'd never require it. This is how you can design systems that work with multiple logins. Let users pick.

To be frank, Google's security is much more advanced than ours. Everything from allowing local IT admins to set password policies, to their hashing algorithms, to supporting Multi-factor auth. It's not unreasonable to want to use Google's login.

> Oh foo, you mean kids would have to use their Google account as their Snap! user name? This is 
the straw that breaks the camel's back, for me, I think.

Not necessarily. We could work around that -- and we would, I believe. All we'd do is ask new accounts to claim a Snap<em>!</em> username. Easy.

The thing is, Google auth limits us storing kids passwords, and that's also a Good Thing. 

To be clear, if we were to support this, I don't think we'd ever go "Google Only". This would mean, users could have both a Snap! account and a Google one. Or one or the other. 

> Are email addresses unique across Snap! accounts?

No. Not at all. This makes me very 😿. But, that won't stop us from making this work. 

The way this typically works is like so:
* if you sign in with Google and it's a new email, we would ask you to pick a Snap<em>!</em> username.
* if that username is taken, you would enter the password and the accounts are linked.
* if it's not taken, you now have a Snap<em>!</em> account with no password, where you need your Google account to sign in. 

It's also worth noting that we currently log IP addresses.

**Comment by bromagosa** [comment]
> To be frank, Google's security is much more advanced than ours.

Except that we can't read their source code to confirm that. They could be sending your data straight to the FBI for all we know :)

My reservations are that if we're supporting Google sign in but not (for example) Twitter sign in, we're effectively favoring one multi-billion company over other ways to sign in. If some kid hasn't made up their mind about getting a Google account because they've heard they can (and will) sell your data, seeing that their favorite programming language endorses signing in with a Google account may tip the scale.

I'm not saying we don't do it, and if we do I like Michael's three-part process. After all, I have a GMail account that I registered in 2005 because this tiny search company called Google was giving away "free" email accounts. Who knew, huh?

**Comment by cycomachead** [comment]
> Regardless, I think we would have to prompt new Google Sign-In users with the ability to set their birth date to complete the signup flow as that data will not be exposed via their Google profile.

Also, my understanding is that is we might not need to do this either. The birthday thing is primarily about COPPA, which Google complies with and has a whole separate system of compliance for Google Apps For Education (GAFE) accounts. 

Which also means from a privacy and data perspective, your GAFE accounts are going to have very different collection policies than regular Google accounts. (However, we as a consumer of the OAuth API wouldn't know which is which type of account. We'd just have to trust Google is doing what they say they are, which is collecting less data about students.)

**Comment by cycomachead** [comment]
> My reservations are that if we're supporting Google sign in but not (for example) Twitter sign in, we're effectively favoring one multi-billion company over other ways to sign in. If some kid hasn't made up their mind about getting a Google account because they've heard they can (and will) sell your data, seeing that their favorite programming language endorses signing in with a Google account may tip the scale.

Yeah, I was just about to say -- we could put our own warning in about our personal beliefs if we wanted. And we can also do it in a way that doesn't encourage signing up for new Google accounts, but says "Already have a google account?" or such so that it's implied that if you need a new account, you should use Snap!'s first.



> Except that we can't read their source code to confirm that. They could be sending your data straight to the FBI for all we know :)

Yeah, that's true. I was mostly thinking on a password level. Google is definitely incentivized to protect their passwords well. And to some extent, using a respected 3rd party auth gives you a little bit of protection against breaches. 

Also, minor semantic thing:

> We don't store any personal information at all,

I think we should remember that we _do_ store identifying info! Email addresses are identifying, and given enough body of programming projects, you can identify people that way. Plus, with the rest of the social site, there will be room for people to give us their info.

Of course, we aren't going to do evil things with people's data! That's the real difference.

And we should be clear too: When a user uses Google Auth, they don't give google much additional info compared to using a Snap<em>!</em> account. Namely, they explicitly tell Google that they're using Snap<em>!</em>, but chances are they got to the Snap! page by Googling for it long ago and the cat was already out of the bag...

**Comment by namoscato** [comment]
Great stuff, @cycomachead – thanks so much for the feedback!

Let me know what you guys think are next steps. As I mentioned earlier, I am certainly willing to help out with this initiative.

A couple of things come to mind:

1. It might be difficult or impossible to work the client button into Snap!'s `<canvas>`. The new beta social site would be nice, but I am not sure what the timeline looks like on that.
2. We might consider keeping the data model and API interface somewhat generic, enabling us to support other providers in the future.

**Comment by cycomachead** [comment]
I like reading things and answering stuff out of order. ;) 

> Does it work like an LDAP server? By which I guess I mean, can we use it without telling Google who we are, what the kid's IP address is, etc.?

Also to add to this example: An LDAP server connected to Snap! would know what Snap! is and would also see the user's IP address, in a default setup. (Although, you could proxy credentials, but that's kinda frowned upon because users would need to trust us with their LDAP creds, which they shouldn't.)

**Comment by cycomachead** [comment]
To the implementation aspect, I guess that’s mostly Bernat’s call. I wouldn’t say this is high priority, but it seem like something we’d use. 

As far as details, handling Snap! login is the last piece, and I’m not yet sure it’s worth doing. (Explanation later). Here’s what I would do

- an identities table that links a Snap! account to an external system, which would have a type or name or something. (Right now that would just be Google)
- handle the login API with dev credentials 
- I would build the login UI all from the social site first. And get that working. 
That’s probably enough to use this. 

- Snap! is special. OAuth requires a redirect, which is a very painful thing to handle in Snap! because Snap is slow to load, and because you cannot reload the page when there is unsaved data. We will need to find a way such that the OAuth redirect would succeed by taking you to a new tab, and then somehow checking the cookie or something. I’m not sure what exactly this will require. 

In general, my believe is that we shouldn’t over abstract a system for multiple providers but we shouldn’t make it hard to adapt in the future. 

Quite honestly, I don’t see I us supporting other OAuth providers BUT I do see supporting other educational protocols like LTI. And if some schools want to pay us money, we could even support CAS or SAML which would let students use their own credentials. 

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by namoscato** [comment]
Sounds good, @cycomachead! I'll see if I can get something started in the coming weeks.
